### ISSUE TITLE
`run-pass` tests should be disabled for non-executable producing targets

### ISSUE THREAD

**Issue by MasterAwesome** [rust-lang/rust#105719]
Currently run-pass tests are enabled for non-executable producing targets. This results in a lot of expected failures when running `./x.py test` and forces us to use `./x.py test --run never` for these targets. IMO this should be the default behavior.

Here's a sample target-spec based on `aarch64-unknown-none` with `"executables": false` set.

```json
{
  "arch": "aarch64",
  "data-layout": "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128",
  "disable-redzone": true,
  "features": "+strict-align,+neon,+fp-armv8",
  "is-builtin": true,
  "linker": "rust-lld",
  "linker-flavor": "ld.lld",
  "llvm-target": "aarch64-unknown-none",
  "max-atomic-width": 128,
  "panic-strategy": "abort",
  "relocation-model": "static",
  "target-pointer-width": "64",
  "executables": false
}
```

I expected to see this happen: *`run-pass` tests are ignored and failing tests specific for this target should be presented.*

Instead, this happened: *Static libraries are generated for `run-pass` tests and test fails for all of them saying `fn main()` is never used and causes test failures due to `dead_code` lint. This is an expected failure since we don't ever want to generate executables with main for this target.*

Possible solution: `compiletest` should detect that a target cannot produce executables and skip `run-pass` tests by default. i.e `./x.py test` should expand to `./x.py test --run never` for targets that *CANNOT* produce executables.
